File: python/src/yosina/transliterators/prolonged_sound_marks.py
"""Prolonged sound marks transliterator."""

from __future__ import annotations

from collections.abc import Iterable
from enum import IntFlag

from ..chars import Char

__all__ = ["Transliterator"]


class CharType(IntFlag):
    """Character type classification flags."""

    OTHER = 0x00
    HIRAGANA = 0x20
    KATAKANA = 0x40
    ALPHABET = 0x60
    DIGIT = 0x80
    EITHER = 0xA0

    # Additional flags
    HALFWIDTH = 1 << 0
    VOWEL_ENDED = 1 << 1
    HATSUON = 1 << 2
    SOKUON = 1 << 3
    PROLONGED_SOUND_MARK = 1 << 4

    # Combined types
    HALFWIDTH_DIGIT = DIGIT | HALFWIDTH
    FULLWIDTH_DIGIT = DIGIT
    HALFWIDTH_ALPHABET = ALPHABET | HALFWIDTH
    FULLWIDTH_ALPHABET = ALPHABET
    ORDINARY_HIRAGANA = HIRAGANA | VOWEL_ENDED
    ORDINARY_KATAKANA = KATAKANA | VOWEL_ENDED
    ORDINARY_HALFWIDTH_KATAKANA = KATAKANA | VOWEL_ENDED | HALFWIDTH

    def is_alnum(self) -> bool:
        """Check if character type is alphanumeric."""
        t = self & 0xE0
        return t == CharType.ALPHABET or t == CharType.DIGIT

    def is_halfwidth(self) -> bool:
        """Check if character type is halfwidth."""
        return bool(self & CharType.HALFWIDTH)

    @classmethod
    def from_codepoint(cls, codepoint: int) -> CharType:
        """Get the character type for a given Unicode codepoint."""
        # Halfwidth digits
        if 0x30 <= codepoint <= 0x39:
            return CharType.HALFWIDTH_DIGIT

        # Fullwidth digits
        if 0xFF10 <= codepoint <= 0xFF19:
            return CharType.FULLWIDTH_DIGIT

        # Halfwidth alphabets
        if (0x41 <= codepoint <= 0x5A) or (0x61 <= codepoint <= 0x7A):
            return CharType.HALFWIDTH_ALPHABET

        # Fullwidth alphabets
        if (0xFF21 <= codepoint <= 0xFF3A) or (0xFF41 <= codepoint <= 0xFF5A):
            return CharType.FULLWIDTH_ALPHABET

        # Special characters
        if codepoint in SPECIALS:
            return SPECIALS[codepoint]

        # Hiragana
        if (0x3041 <= codepoint <= 0x309C) or codepoint == 0x309F:
            return CharType.ORDINARY_HIRAGANA

        # Katakana
        if (0x30A1 <= codepoint <= 0x30FA) or (0x30FD <= codepoint <= 0x30FF):
            return CharType.ORDINARY_KATAKANA

        # Halfwidth katakana
        if (0xFF66 <= codepoint <= 0xFF6F) or (0xFF71 <= codepoint <= 0xFF9F):
            return CharType.ORDINARY_HALFWIDTH_KATAKANA

        return CharType.OTHER


# Special character mappings
SPECIALS = {
    0xFF70: CharType.KATAKANA | CharType.PROLONGED_SOUND_MARK | CharType.HALFWIDTH,  # ｰ
    0x30FC: CharType.EITHER | CharType.PROLONGED_SOUND_MARK,  # ー
    0x3063: CharType.HIRAGANA | CharType.SOKUON,  # っ
    0x3093: CharType.HIRAGANA | CharType.HATSUON,  # ん
    0x30C3: CharType.KATAKANA | CharType.SOKUON,  # ッ
    0x30F3: CharType.KATAKANA | CharType.HATSUON,  # ン
    0xFF6F: CharType.KATAKANA | CharType.SOKUON | CharType.HALFWIDTH,  # ｯ
    0xFF9D: CharType.KATAKANA | CharType.HATSUON | CharType.HALFWIDTH,  # ﾝ
}

# Hyphen-like characters that could be prolonged sound marks
HYPHEN_LIKE_CHARS = {
    "\u002d",
    "\u2010",
    "\u2014",
    "\u2015",
    "\u2212",
    "\uff0d",
    "\uff70",
    "\u30fc",
}


class Transliterator:
    """Prolonged sound marks transliterator.

    This transliterator handles the replacement of hyphen-like characters with
    appropriate prolonged sound marks (ー or ｰ) when they appear after Japanese
    characters that can be prolonged.
    """

    skip_already_transliterated_chars: bool
    allow_prolonged_hatsuon: bool
    allow_prolonged_sokuon: bool
    replace_prolonged_marks_following_alnums: bool
    prolongables: CharType

    def __init__(
        self,
        *,
        skip_already_transliterated_chars: bool = False,
        allow_prolonged_hatsuon: bool = False,
        allow_prolonged_sokuon: bool = False,
        replace_prolonged_marks_following_alnums: bool = False,
    ) -> None:
        """Initialize the transliterator with options.

        :param skip_already_transliterated_chars: Skip chars that have been transliterated. Defaults to False.
        :param allow_prolonged_hatsuon: Allow prolonging hatsuon characters (ん/ン). Defaults to False.
        :param allow_prolonged_sokuon: Allow prolonging sokuon characters (っ/ッ). Defaults to False.
        :param replace_prolonged_marks_following_alnums: Whether to replace prolonged voice marks following an
               alphanumeric. Defaults to False.
        """
        self.skip_already_transliterated_chars = skip_already_transliterated_chars
        self.allow_prolonged_hatsuon = allow_prolonged_hatsuon
        self.allow_prolonged_sokuon = allow_prolonged_sokuon
        self.replace_prolonged_marks_following_alnums = replace_prolonged_marks_following_alnums

        # Build prolongable character types
        self.prolongables = CharType.VOWEL_ENDED | CharType.PROLONGED_SOUND_MARK
        if allow_prolonged_hatsuon:
            self.prolongables |= CharType.HATSUON
        if allow_prolonged_sokuon:
            self.prolongables |= CharType.SOKUON

    def __call__(self, input_chars: Iterable[Char]) -> Iterable[Char]:
        """Transliterate prolonged sound marks."""
        offset = 0
        processed_chars_in_lookahead = False
        lookahead_buf: list[Char] = []
        last_non_prolonged_char: tuple[Char, CharType] | None = None

        for char in input_chars:
            if lookahead_buf:
                if char.c in HYPHEN_LIKE_CHARS:
                    if char.source is not None:
                        processed_chars_in_lookahead = True
                    lookahead_buf.append(char)
                    continue
                # Process buffered characters
                prev_non_prolonged_char = last_non_prolonged_char
                codepoint = ord(char.c[0]) if char.c else -1
                last_non_prolonged_char = (char, CharType.from_codepoint(codepoint))

                # Check if we should replace with hyphens for alphanumerics
                if (prev_non_prolonged_char is None or prev_non_prolonged_char[1].is_alnum()) and (
                    not self.skip_already_transliterated_chars or not processed_chars_in_lookahead
                ):
                    if (
                        last_non_prolonged_char[1].is_halfwidth()
                        if prev_non_prolonged_char is None
                        else prev_non_prolonged_char[1].is_halfwidth()
                    ):
                        replacement = "\u002d"
                    else:
                        replacement = "\uff0d"

                    for buffered_char in lookahead_buf:
                        yield Char(c=replacement, offset=offset, source=buffered_char)
                        offset += len(replacement)
                else:
                    # Just pass through the buffered characters
                    for buffered_char in lookahead_buf:
                        yield buffered_char.with_offset(offset)
                        offset += len(buffered_char.c)

                lookahead_buf = []
                yield char.with_offset(offset)
                offset += len(char.c)
                codepoint = ord(char.c[0]) if char.c else -1
                last_non_prolonged_char = (char, CharType.from_codepoint(codepoint))
                processed_chars_in_lookahead = False
                continue

            # Check if this is a hyphen-like character that might be a prolonged sound mark
            if char.c in HYPHEN_LIKE_CHARS:
                should_process = not self.skip_already_transliterated_chars or not char.is_transliterated()
                if should_process and last_non_prolonged_char is not None:
                    if self.prolongables & last_non_prolonged_char[1]:
                        replacement = "\uff70" if last_non_prolonged_char[1].is_halfwidth() else "\u30fc"
                        yield Char(c=replacement, offset=offset, source=char)
                        offset += len(replacement)
                        continue
                    else:
                        # Check if we should buffer for alphanumeric replacement
                        if self.replace_prolonged_marks_following_alnums and last_non_prolonged_char[1].is_alnum():
                            lookahead_buf.append(char)
                            continue
            else:
                # Update last non-prolonged character
                codepoint = ord(char.c[0]) if char.c else -1
                last_non_prolonged_char = (char, CharType.from_codepoint(codepoint))

            # Default: pass through the character
            yield char.with_offset(offset)
            offset += len(char.c)

File: ruby/lib/yosina/transliterators/prolonged_sound_marks.rb
# frozen_string_literal: true

module Yosina
  module Transliterators
    # Handle prolonged sound marks transliterator
    module ProlongedSoundMarks
      # Hyphen-like characters that can be converted to prolonged sound marks
      HYPHEN_LIKE_CHARS = [
        "\u002d",  # HYPHEN-MINUS
        "\u2010",  # HYPHEN
        "\u2014",  # EM DASH
        "\u2015",  # HORIZONTAL BAR
        "\u2212",  # MINUS SIGN
        "\uff0d",  # FULLWIDTH HYPHEN-MINUS
        "\uff70",  # HALFWIDTH KATAKANA-HIRAGANA PROLONGED SOUND MARK (already converted)
        "\u30fc"   # KATAKANA-HIRAGANA PROLONGED SOUND MARK (already converted)
      ].freeze

      # Fullwidth prolonged sound mark
      FULLWIDTH_PROLONGED_MARK = "\u30fc" # ー

      # Halfwidth prolonged sound mark
      HALFWIDTH_PROLONGED_MARK = "\uff70" # ｰ

      # Hatsuon characters (ん/ン/ﾝ)
      HATSUON_CHARS = %w[ん ン ﾝ].freeze

      # Sokuon characters (っ/ッ/ｯ)
      SOKUON_CHARS = %w[っ ッ ｯ].freeze

      # Characters that can be prolonged by default:
      # - All hiragana characters (U+3041-U+309C, U+309F) except hatsuon/sokuon
      # - All katakana characters (U+30A1-U+30FA, U+30FD-U+30FF) except hatsuon/sokuon
      # - All halfwidth katakana characters (U+FF66-U+FF6F, U+FF71-U+FF9F) except hatsuon/sokuon
      # - Prolonged sound marks themselves (ー, ｰ)
      # This matches the behavior in other language implementations where all
      # Japanese characters are considered "vowel-ended" and thus prolongable
      def self.default_prolongable_chars
        chars = []

        # Hiragana range (U+3041-U+309C, U+309F)
        (0x3041..0x309C).each { |cp| chars << cp.chr('UTF-8') }
        chars << 0x309F.chr('UTF-8')

        # Katakana range (U+30A1-U+30FA, U+30FD-U+30FF)
        (0x30A1..0x30FA).each { |cp| chars << cp.chr('UTF-8') }
        (0x30FD..0x30FF).each { |cp| chars << cp.chr('UTF-8') }

        # Halfwidth katakana range (U+FF66-U+FF6F, U+FF71-U+FF9F)
        (0xFF66..0xFF6F).each { |cp| chars << cp.chr('UTF-8') }
        (0xFF71..0xFF9F).each { |cp| chars << cp.chr('UTF-8') }

        # Add prolonged sound marks explicitly
        chars << FULLWIDTH_PROLONGED_MARK  # ー
        chars << HALFWIDTH_PROLONGED_MARK  # ｰ

        # Remove special characters (hatsuon, sokuon) as they're handled separately
        chars - HATSUON_CHARS - SOKUON_CHARS
      end

      # Cache the result since it's expensive to compute each time
      DEFAULT_PROLONGABLE_CHARS = default_prolongable_chars.freeze

      # Check if character is halfwidth Japanese
      def self.halfwidth_japanese?(char)
        # Halfwidth katakana range U+FF65-U+FF9F
        char_code = char.ord
        char_code >= 0xFF65 && char_code <= 0xFF9F
      end

      # Check if character is fullwidth Japanese
      def self.fullwidth_japanese?(char)
        char_code = char.ord
        # Hiragana: U+3040-U+309F
        # Katakana: U+30A0-U+30FF
        (char_code >= 0x3040 && char_code <= 0x309F) ||
          (char_code >= 0x30A0 && char_code <= 0x30FF)
      end

      # Check if character is alphanumeric
      def self.alphanumeric?(char)
        char.match?(/[a-zA-Z0-9]/)
      end

      # Check if character is fullwidth alphanumeric
      def self.fullwidth_alphanumeric?(char)
        char_code = char.ord
        # Fullwidth ASCII: U+FF00-U+FF5E
        char_code >= 0xFF00 && char_code <= 0xFF5E
      end

      # Transliterator for prolonged sound marks
      class Transliterator < Yosina::BaseTransliterator
        attr_reader :skip_already_transliterated_chars, :allow_prolonged_hatsuon,
                    :allow_prolonged_sokuon, :replace_prolonged_marks_following_alnums

        # Initialize the transliterator with options
        #
        # @param options [Hash] Configuration options
        # @option options [Boolean] :skip_already_transliterated_chars Skip chars that were already processed.
        #     Default: false.
        # @option options [Boolean] :allow_prolonged_hatsuon Allow prolonging ん/ン/ﾝ.
        #     Default: false.
        # @option options [Boolean] :allow_prolonged_sokuon Allow prolonging っ/ッ/ｯ.
        #     Default: false.
        # @option options [Boolean] :replace_prolonged_marks_following_alnums Replace prolonged marks after alphanum
        #    with hyphens. Default: false.
        def initialize(options = {})
          super()
          @skip_already_transliterated_chars = options.fetch(:skip_already_transliterated_chars, false)
          @allow_prolonged_hatsuon = options.fetch(:allow_prolonged_hatsuon, false)
          @allow_prolonged_sokuon = options.fetch(:allow_prolonged_sokuon, false)
          @replace_prolonged_marks_following_alnums = options.fetch(:replace_prolonged_marks_following_alnums, false)
        end

        # Convert hyphen-like characters to appropriate prolonged sound marks
        #
        # @param input_chars [Array<Char>] The characters to transliterate
        # @return [Array<Char>] The transliterated characters
        def call(input_chars)
          offset = 0
          result = []
          i = 0

          while i < input_chars.length
            char = input_chars[i]

            # Skip if we should skip already transliterated chars and this char was processed
            if @skip_already_transliterated_chars && char.source
              result << Char.new(c: char.c, offset: offset, source: char.source)
              offset += char.c.length
              i += 1
              next
            end

            # Check if current character is a hyphen-like character
            if HYPHEN_LIKE_CHARS.include?(char.c)
              replacement = get_prolonged_mark_replacement(result.last)

              if replacement
                result << Char.new(c: replacement, offset: offset, source: char)
                offset += replacement.length
              else
                # No replacement, keep original
                result << Char.new(c: char.c, offset: offset, source: char.source)
                offset += char.c.length
              end
            else
              # Not a hyphen-like character, keep as is
              result << Char.new(c: char.c, offset: offset, source: char.source)
              offset += char.c.length
            end

            i += 1
          end

          result
        end

        private

        # Get replacement for hyphen-like character based on context
        #
        # @param prev_char [Char, nil] The previous character for context
        # @return [String, nil] The replacement character or nil if no replacement
        def get_prolonged_mark_replacement(prev_char)
          # If no previous character, can't determine context
          return nil unless prev_char

          prev_char_str = prev_char.c

          # Check if previous character is prolongable
          if prolongable_char?(prev_char_str)
            # Determine appropriate prolonged mark based on character width
            if ProlongedSoundMarks.halfwidth_japanese?(prev_char_str)
              return HALFWIDTH_PROLONGED_MARK
            elsif ProlongedSoundMarks.fullwidth_japanese?(prev_char_str)
              return FULLWIDTH_PROLONGED_MARK
            end
          end

          # Handle alphanumeric replacement option
          if @replace_prolonged_marks_following_alnums
            if ProlongedSoundMarks.alphanumeric?(prev_char_str)
              return "\u002d"  # Regular hyphen
            elsif ProlongedSoundMarks.fullwidth_alphanumeric?(prev_char_str)
              return "\uff0d"  # Fullwidth hyphen
            end
          end

          # No replacement appropriate
          nil
        end

        # Check if a character can be prolonged
        #
        # @param char [String] The character to check
        # @return [Boolean] True if the character can be prolonged
        def prolongable_char?(char)
          # Default prolongable characters
          return true if DEFAULT_PROLONGABLE_CHARS.include?(char)

          # Hatsuon (if allowed)
          return true if @allow_prolonged_hatsuon && HATSUON_CHARS.include?(char)

          # Sokuon (if allowed)
          return true if @allow_prolonged_sokuon && SOKUON_CHARS.include?(char)

          false
        end
      end

      # Factory method to create a prolonged sound marks transliterator
      #
      # @param options [Hash] Configuration options
      # @return [Transliterator] A new prolonged sound marks transliterator instance
      def self.call(options = {})
        Transliterator.new(options)
      end
    end
  end
end
